extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/dart/controller/sql/DartQueryMaker.java

```diff
@@ -0,0 +1,466 @@
+  )
+  {
+    this.fieldMapping = fieldMapping;
+    this.controllerContext = controllerContext;
+    this.plannerContext = plannerContext;
+    this.controllerRegistry = controllerRegistry;
+    this.controllerConfig = controllerConfig;
+    this.controllerExecutor = controllerExecutor;
+  }
+
+  @Override
+  public QueryResponse<Object[]> runQuery(DruidQuery druidQuery)
+  {
+    final MSQSpec querySpec = MSQTaskQueryMaker.makeQuerySpec(
+        null,
+        druidQuery,
+        fieldMapping,
+        plannerContext,
+        null // Only used for DML, which this isn't
+    );
+    final List<Pair<SqlTypeName, ColumnType>> types =
+        MSQTaskQueryMaker.getTypes(druidQuery, fieldMapping, plannerContext);
+
+    final ControllerImpl controller = new ControllerImpl(
+        druidQuery.getQuery().context().getString(DartSqlEngine.CTX_DART_QUERY_ID),
+        querySpec,
+        new ResultsContext(
+            types.stream().map(p -> p.lhs).collect(Collectors.toList()),
+            SqlResults.Context.fromPlannerContext(plannerContext)
+        ),
+        controllerContext
+    );
+
+    final ControllerHolder controllerHolder = new ControllerHolder(
+        controller,
+        plannerContext.getSqlQueryId(),
+        plannerContext.getSql(),
+        plannerContext.getAuthenticationResult(),
+        DateTimes.nowUtc()
+    );
+
+    final boolean fullReport = druidQuery.getQuery().context().getBoolean(
+        DartSqlEngine.CTX_FULL_REPORT,
+        DartSqlEngine.CTX_FULL_REPORT_DEFAULT
+    );
+
+    // Register controller before submitting anything to controllerExeuctor, so it shows up in
+    // "active controllers" lists.
+    controllerRegistry.register(controllerHolder);
+
+    try {
+      final Sequence<Object[]> results =
+          fullReport ? runWithReport(controllerHolder) : runWithoutReport(controllerHolder);
+      return QueryResponse.withEmptyContext(results);
+    }
+    catch (Throwable e) {
+      // Error while creating the result sequence; unregister controller.
+      controllerRegistry.remove(controllerHolder);
+      throw e;
+    }
+  }
+
+  /**
+   * Run a query and return the full report, buffered in memory up to
+   * {@link DartControllerConfig#getMaxQueryReportSize()}.
+   *
+   * Arranges for {@link DartControllerRegistry#remove(ControllerHolder)} to be called upon completion (either
+   * success or failure).
+   */
+  private Sequence<Object[]> runWithReport(final ControllerHolder controllerHolder)
+  {
+    final Future<Map<String, Object>> reportFuture;
+
+    // Run in controllerExecutor. Control doesn't really *need* to be moved to another thread, but we have to
+    // use the controllerExecutor anyway, to ensure we respect the concurrentQueries configuration.
+    reportFuture = controllerExecutor.submit(() -> {
+      try {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        final TaskReportQueryListener queryListener = new TaskReportQueryListener(
```

Comment:
Hmm. It's still writing a thing called a `MSQTaskReport` that extends `TaskReport`. So I think the name makes sense. However, it might make sense to define some common object that is used by the `MSQTaskReport` container and the Dart API. If it's ok I'd like to leave this thought for the future.